Kernel/Process.cpp

```diff
@@ -899,23 +930,138 @@ KResultOr<Vector<String>> Process::find_shebang_interpreter_for_executable(const
+                m_name.characters(), m_pid, path.characters(), interpreter_description->absolute_path().characters(), interpreter_interpreter_path.characters());
+            return KResult(-ELOOP);
+        }
+
+        return interpreter_description;
+    }
+
+    if (elf_header->e_type != ET_EXEC) {
+        // We can't exec an ET_REL, that's just an object file from the compiler
+        // If it's ET_DYN with no PT_INTERP, then we can't load it properly either
+        return KResult(-ENOEXEC);
+    }
+
+    // No interpreter, but, path refers to a valid elf image
+    return KResult(KSuccess);
+}
+
+int Process::exec(String path, Vector<String> arguments, Vector<String> environment, int recursion_depth)
 {
-    auto result = find_shebang_interpreter_for_executable(path);
-    if (!result.is_error()) {
-        Vector<String> new_arguments(result.value());
+    if (recursion_depth > 2) {
+        dbgprintf("%s(%d) exec(%s): SHENANIGANS! recursed too far trying to find #! interpreter\n", m_name.characters(), m_pid, path.characters());
+        return -ELOOP;
+    }
+
+    // Open the file to check what kind of binary format it is
+    // Currently supported formats:
+    //    - #! interpreted file
+    //    - ELF32
+    //        * ET_EXEC binary that just gets loaded
+    //        * ET_DYN binary that requires a program interpreter
+    //
+    auto result = VFS::the().open(path, O_EXEC, 0, current_directory());
+    if (result.is_error())
+        return result.error();
+    auto description = result.value();
+    auto metadata = description->metadata();
+
+    // Always gonna need at least 3 bytes. these are for #!X
+    if (metadata.size < 3)
+        return -ENOEXEC;
+
+    ASSERT(description->inode());
+
+    // Read the first page of the program into memory so we can validate the binfmt of it
+    char first_page[PAGE_SIZE];
+    int nread = description->read((u8*)&first_page, sizeof(first_page));
+
+    // 1) #! interpreted file
+    auto shebang_result = find_shebang_interpreter_for_executable(first_page, nread);
+    if (!shebang_result.is_error()) {
+        Vector<String> new_arguments(shebang_result.value());
 
         new_arguments.append(path);
 
         arguments.remove(0);
         new_arguments.append(move(arguments));
 
-        return exec(result.value().first(), move(new_arguments), move(environment));
+        return exec(shebang_result.value().first(), move(new_arguments), move(environment), ++recursion_depth);
     }
 
+    // #2) ELF32 for i386
+    auto elf_result = find_elf_interpreter_for_executable(path, first_page, nread, metadata.size);
+    RefPtr<FileDescription> interpreter_description;
+    // We're getting either an interpreter, an error, or KSuccess (i.e. no interpreter but file checks out)
+    if (!elf_result.is_error())
+        interpreter_description = elf_result.value();
+    else if (elf_result.error().is_error())
+        return elf_result.error();
```

Comment:
This reads quite poorly. "If the error is an error", ehhh. :)
I understand the predicament here though, since `KSuccess` is a valid return value in this case.